### PR TITLE
halo: init at 2.15.1

### PR DESCRIPTION
**Pull request by wangxiaoerYah**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by wangxiaoerYah** [comment]
This is a self-hosted blog program that uses java language.

Official website: https://www.halo.run/

Usage example: `HALO_WORK_DIR=~/halo halo --server.port=9000 --halo.external-url=http://localhost:9000/`

File: pkgs/servers/halo/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{ pkgs
```
Comment (wineee) [inline pkgs/servers/halo/default.nix:1]: remove this file

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -1876,6 +1876,8 @@ with pkgs;
 
   fwbuilder = libsForQt5.callPackage ../tools/security/fwbuilder { };
 
+  halo = callPackage ../servers/halo { };
```
Comment (wineee) [inline pkgs/top-level/all-packages.nix:1879]:
```suggestion
```

**Review by wineee — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ha/halo/package.nix [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  pname = "halo";
+  version = "2.15.1";
+  src = fetchurl {
+    url = "https://github.com/halo-dev/halo/releases/download/v${version}/${pname}-${version}.jar";
+    hash = "sha256-6qgbUXd0E38Rhv0Jf1qP54yJhQiB1a/Jr1FYw3bC4qI=";
+  };
+
+  nativeBuildInputs = [
+    makeWrapper
+    pkgs.rsync
```
Comment (wineee) [inline pkgs/by-name/ha/halo/package.nix:19]:
```suggestion
    rsync
```

File: pkgs/by-name/ha/halo/package.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{ pkgs
+, lib
+, stdenv
+, fetchurl
+, makeWrapper
+, temurin-jre-bin
+, ...
+}:
```
Comment (wineee) [inline pkgs/by-name/ha/halo/package.nix:8]:
```suggestion
{ lib
, stdenv
, fetchurl
, makeWrapper
, temurin-jre-bin
, rsync
}:
```

Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:8]: thank you

**Comment by wineee** [comment]
following https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions

Commit messages should be `halo: init at 2.15.1`

**Comment by wangxiaoerYah** [comment]
> halo: init at 2.15.1

Correct it immediately. Thanks for the guidance.

File: pkgs/by-name/ha/halo/package.nix [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  ];
+
+  buildInputs = [
+    temurin-jre-bin
+  ];
+
+  dontPatch = true;
+  dontBuild = true;
+  dontConfigure = true;
+  dontPatchELF = true;
```
Comment (wineee) [inline pkgs/by-name/ha/halo/package.nix:29]: You are not using autoPatchelfHook， don't need set `dontPatchELF`

File: pkgs/by-name/ha/halo/package.nix [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  noAuditTmpdir = true;
```
Comment (wineee) [inline pkgs/by-name/ha/halo/package.nix:30]: why we need `noAuditTmpdir` ?

**Review by wineee — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ha/halo/package.nix (hunk 6 above)
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:30]: Change it immediately. It was originally to reduce the build steps and speed up debugging.

**Review by zzzsyyy — COMMENTED** [review]
you should follow [here](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md#individual-maintainer-list) to add yourself to the maintainers list 

and please squash the commits, it should be 
+ `maintainers: add xxx`
+ `hola: init at 2.15.1`

File: pkgs/by-name/ha/halo/package.nix [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+      --add-flags "org.springframework.boot.loader.launch.JarLauncher"
+  '';
+
+  meta = {
+    homepage = "https://www.halo.run";
+    description = "Self-hosted dynamic blogging program";
+    maintainers = with lib.maintainers; [ yah ];
+    license = lib.licenses.gpl3;
+    platforms = [ "aarch64-linux" "x86_64-linux" ];
+    broken = (stdenv.buildPlatform != stdenv.hostPlatform);
```
Comment (linsui) [inline pkgs/by-name/ha/halo/package.nix:58]: What's this?

File: pkgs/by-name/ha/halo/package.nix [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      --set SPRING_CONFIG_LOCATION "optional:classpath:/;optional:file:\`\$HALO_WORK_DIR\`/" \
+      --add-flags "-server \$JVM_OPTS" \
+      --add-flags "org.springframework.boot.loader.launch.JarLauncher"
+  '';
+
+  meta = {
+    homepage = "https://www.halo.run";
+    description = "Self-hosted dynamic blogging program";
+    maintainers = with lib.maintainers; [ yah ];
+    license = lib.licenses.gpl3;
```
Comment (linsui) [inline pkgs/by-name/ha/halo/package.nix:56]:
```suggestion
    license = lib.licenses.gpl3Only;
```

File: pkgs/by-name/ha/halo/package.nix [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+  nativeBuildInputs = [
+    makeWrapper
+    rsync
+  ];
+
+  buildInputs = [
+    temurin-jre-bin
+  ];
+
```
Comment (linsui) [inline pkgs/by-name/ha/halo/package.nix:25]:
```suggestion
```

File: pkgs/by-name/ha/halo/package.nix [hunk 10]
```diff
@@ -0,0 +1,61 @@
+{ lib
+, stdenv
+, fetchurl
+, makeWrapper
+, temurin-jre-bin
```
Comment (linsui) [inline pkgs/by-name/ha/halo/package.nix:5]: Why do you use this jre install of default jdk?

File: pkgs/by-name/ha/halo/package.nix [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  dontConfigure = true;
+
+  unpackPhase = ''
+    cp $src ${pname}.jar
+    java -Djarmode=tools -jar ${pname}.jar extract --layers --launcher
+  '';
+
+  installPhase = ''
+    mkdir -p $out/application
+    rsync -a ${pname}/dependencies/ $out/application
```
Comment (linsui) [inline pkgs/by-name/ha/halo/package.nix:37]: Why do you use rsync instead of mv/cp?

**Comment by linsui** [comment]
Can you build the jar from source?

File: pkgs/by-name/ha/halo/package.nix (hunk 9 above)
Comment (linsui) [reply pkgs/by-name/ha/halo/package.nix:25]: This should be put in nativeBuildInputs.

Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:25]: ok i got

File: pkgs/by-name/ha/halo/package.nix (hunk 10 above)
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:5]: Because the official is using eclipse's jre. I want to follow the official choice.

**Comment by wangxiaoerYah** [comment]
> 你能从源代码构建 jar 吗？



> Can you build the jar from source?

Of course, do I need to add a test step? I use jar because the official release of the relase version is directly available. If I compile it again, it may waste the use of CI build.

Do I need to build from scratch? If so, I will submit it later after successful testing.

File: pkgs/by-name/ha/halo/package.nix (hunk 10 above)
Comment (linsui) [reply pkgs/by-name/ha/halo/package.nix:5]: There is no official openjdk binary so it's reasonable to use temurin jdk upstream. I thought it's better to use the default jdk in distribution unless it doesn't work.

File: pkgs/by-name/ha/halo/package.nix [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      --add-flags "org.springframework.boot.loader.launch.JarLauncher"
+  '';
+
+  meta = {
+    homepage = "https://www.halo.run";
+    description = "Self-hosted dynamic blogging program";
+    maintainers = with lib.maintainers; [ yah ];
+    license = lib.licenses.gpl3Only;
+    platforms = [ "aarch64-linux" "x86_64-linux" ];
+    mainProgram = "halo";
```
[the remark this answers is not in the document]
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:55]: Thank you 🙏, I will change it immediately.

**Comment by linsui** [comment]
It's not required to build it from source but it's preferred.

File: pkgs/by-name/ha/halo/package.nix [hunk 13]
```diff
@@ -0,0 +1,57 @@
+{ lib
+, stdenv
+, fetchurl
+, makeWrapper
+, temurin-jre-bin
+, rsync
+, ...
```
[the remark this answers is not in the document]
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:7]: Thank you 🙏, I will change it immediately.

File: pkgs/by-name/ha/halo/package.nix (hunk 7 above)
Comment (linsui) [reply pkgs/by-name/ha/halo/package.nix]: I thought it's enough to set platforms.

File: pkgs/by-name/ha/halo/package.nix [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+    mkdir -p $out/bin
+    makeWrapper ${temurin-jre-bin}/bin/java $out/bin/${pname} \
+      --chdir $out/application \
+      --set-default HALO_WORK_DIR "/var/lib/${pname}" \
+      --set-default JVM_OPTS "-Xms256m -Xmx256m" \
+      --set SPRING_CONFIG_LOCATION "optional:classpath:/;optional:file:\`\$HALO_WORK_DIR\`/" \
+      --add-flags "-server \$JVM_OPTS" \
+      --add-flags "org.springframework.boot.loader.launch.JarLauncher"
+  '';
```
[the remark this answers is not in the document]
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:47]: What does it mean here? Do I need to set it even if there is no pre-step? Or do I need to split it out?

[the remark this answers is not in the document]
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:47]: Got it, this explanation is perfect. Thank you.

File: pkgs/by-name/ha/halo/package.nix (hunk 10 above)
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix:5]: I actually want to use the smallest version, I have tried jre_mini but it doesn't come with spring.
I will try again with openjdk.

File: pkgs/by-name/ha/halo/package.nix [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    rsync -a ${pname}/application/ $out/application
+
+    mkdir -p $out/bin
+    makeWrapper ${temurin-jre-bin}/bin/java $out/bin/${pname} \
+      --chdir $out/application \
+      --set-default HALO_WORK_DIR "/var/lib/${pname}" \
+      --set-default JVM_OPTS "-Xms256m -Xmx256m" \
+      --set SPRING_CONFIG_LOCATION "optional:classpath:/;optional:file:\`\$HALO_WORK_DIR\`/" \
+      --add-flags "-server \$JVM_OPTS" \
+      --add-flags "org.springframework.boot.loader.launch.JarLauncher"
```
Comment (linsui) [inline pkgs/by-name/ha/halo/package.nix:47]: Not sure if these options should be hardcoded here. It's better to put them in a module.

**Review by Aleksanaa — COMMENTED** [review]
Looks good overall.

**Comment by wineee** [comment]
Please rebase the commits, it should be
- maintainers: add yah
- hola: init at 2.15.1

not
- hola: init at 2.15.1
- maintainers: add yah

File: pkgs/by-name/ha/halo/package.nix [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    url = "https://github.com/halo-dev/halo/releases/download/v${version}/${pname}-${version}.jar";
+    hash = "sha256-6qgbUXd0E38Rhv0Jf1qP54yJhQiB1a/Jr1FYw3bC4qI=";
+  };
+
+  nativeBuildInputs = [
+    makeWrapper
+    temurin-jre-bin
+  ];
+
+  dontPatch = true;
```
Comment (wineee) [inline pkgs/by-name/ha/halo/package.nix:21]: set `dontPatch` should be unnecessary

**Review by wineee — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ha/halo/package.nix (hunk 16 above)
Comment (wangxiaoerYah) [reply pkgs/by-name/ha/halo/package.nix]: ok.

**Review by wineee — APPROVED** [review]

**Pull request merged**